console-subscriber/src/record.rs

```diff
@@ -0,0 +1,224 @@
+/// older versions.
+///
+/// But while this is in rapid development, we can move fast and break things.
+const DATA_FORMAT_VERSION: u8 = 1;
+
+pub(crate) struct Recorder {
+    buf: Arc<Mutex<RecordBuf>>,
+
+    worker: std::thread::JoinHandle<()>,
+}
+
+struct Io {
+    buf: Arc<Mutex<RecordBuf>>,
+    file: File,
+}
+
+struct RecordBuf {
+    /// The current buffer to serialize events into.
+    bytes: Vec<u8>,
+    /// The "next" buffer that should be used when the IO thread takes the
+    /// current buffer. After flushing, the IO thread will put the buffer
+    /// back in this slot, so the allocation can be reused.
+    next: Vec<u8>,
+}
+
+#[derive(Serialize)]
+struct Header {
+    v: u8,
+}
+
+#[derive(Serialize)]
+enum Event<'a> {
+    Spawn {
+        id: u64,
+        at: SystemTime,
+        fields: SerializeFields<'a>,
+    },
+    Enter {
+        id: u64,
+        at: SystemTime,
+    },
+    Exit {
+        id: u64,
+        at: SystemTime,
+    },
+    Close {
+        id: u64,
+        at: SystemTime,
+    },
+    Waker {
+        id: u64,
+        op: super::WakeOp,
+        at: SystemTime,
+    },
+}
+
+struct SerializeFields<'a>(&'a [proto::Field]);
+
+struct SerializeField<'a>(&'a proto::Field);
+
+impl Recorder {
+    pub(crate) fn new(path: &Path) -> io::Result<Self> {
+        let buf = Arc::new(Mutex::new(RecordBuf::new()));
+        let buf2 = buf.clone();
+        let file = std::fs::File::create(path)?;
+
+        let worker = std::thread::Builder::new()
+            .name("console/subscriber/recorder/io".into())
+            .spawn(move || {
+                record_io(Io { buf: buf2, file });
+            })?;
+
+        let recorder = Recorder { buf, worker };
+
+        recorder.write(&Header {
+            v: DATA_FORMAT_VERSION,
+        });
+
+        Ok(recorder)
+    }
+
+    pub(crate) fn record(&self, event: &crate::Event) {
+        let event = match event {
+            crate::Event::Spawn { id, at, fields, .. } => Event::Spawn {
+                id: id.into_u64(),
+                at: *at,
+                fields: SerializeFields(fields),
+            },
+            crate::Event::Enter { id, at } => Event::Enter {
+                id: id.into_u64(),
+                at: *at,
+            },
+            crate::Event::Exit { id, at } => Event::Exit {
+                id: id.into_u64(),
+                at: *at,
+            },
+            crate::Event::Close { id, at } => Event::Close {
+                id: id.into_u64(),
+                at: *at,
+            },
+            crate::Event::Waker { id, op, at } => Event::Waker {
+                id: id.into_u64(),
+                at: *at,
+                op: *op,
+            },
+            _ => return,
+        };
+
+        self.write(&event);
+    }
+
+    fn write<T: Serialize>(&self, val: &T) {
+        let mut buf = self.buf.lock().unwrap();
+        serde_json::to_writer(&mut buf.bytes, val).expect("json");
+        buf.bytes.push(b'\n');
```

Comment:
technically, this means that the output format is not actually JSON, but _newline-delimited_ JSON, where each line is a separate JSON object. this means that (e.g.) `jq` doesn't like the format without some pre-processing so that it's invoked separately on each line.

i wonder if we want to change the output format so that the whole file is a valid JSON object. we could emit something like this instead:
```
{
    "header": <header>,
    "events": [
        <event 1>,
        <event 2>,
        ...
        <event n>
    ]
}
```
that way, the whole file would parse as JSON and could be more easily used with tools like `jq`.

the downside to this is just that we would have to detect shutdowns and ensure that we properly emit the close delimiters for the JSON object...which would introduce more complexity for sure. wdyt?

Reply:
Yup, makes sense!

Reply:
it would maybe be nice for `jq` to just work, but i think since the _primary_ intended use case for these files is to parse them with the console app, optimizing for parsing in slices is a good call.